library/std/src/io/buffered/bufwriter.rs

```diff
@@ -291,64 +374,151 @@ impl<W: Write> BufWriter<W> {
+            let buf = tail!(self.flush_buf_vectored(buf)?);
+
+            // At this point, because we know that buf.len() < self.buf.len(),
+            // and that the buffer has been flushed we know that this will
+            // succeed in totality
+            self.write_to_buf(buf);
         }
-        // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
-        if buf.len() >= self.buf.capacity() {
-            self.panicked = true;
-            let r = self.get_mut().write_all(buf);
-            self.panicked = false;
-            r
-        } else {
-            self.buf.extend_from_slice(buf);
-            Ok(())
+
+        // If, at this point, the buffer is full, we may as well eagerly
+        // attempt to flush, so that the next write will have an empty
+        // buffer.
+        if self.available() == 0 {
+            self.flush_buf()?;
         }
+
+        Ok(())
     }
 
     fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
-        let total_len = bufs.iter().map(|b| b.len()).sum::<usize>();
-        if self.buf.len() + total_len > self.buf.capacity() {
-            self.flush_buf()?;
-        }
-        // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
-        if total_len >= self.buf.capacity() {
-            self.panicked = true;
-            let r = self.get_mut().write_vectored(bufs);
-            self.panicked = false;
-            r
+        if let Some(buf) = only_one(bufs, |b| !b.is_empty()) {
+            // If there's exactly 1 incoming buffer, `Self::write` can make
+            // use of self.inner.write_vectored to attempt to combine flushing
+            // the existing buffer with writing the new one.
+            self.write(buf)
+        } else if self.get_ref().is_write_vectored() {
+            let total_len: usize = bufs.iter().map(|buf| buf.len()).sum();
+
+            if total_len > self.available() {
+                self.flush_buf()?;
+            }
+
+            if total_len >= self.capacity() {
+                self.get_mut().write_vectored(bufs)
+            } else {
+                // Correctness note: we've already verified that none of these
+                // will overflow the buffer, because total_len < capacity
+                bufs.iter().for_each(|buf| self.buf.extend_from_slice(buf));
+                Ok(total_len)
+            }
         } else {
-            bufs.iter().for_each(|b| self.buf.extend_from_slice(b));
-            Ok(total_len)
+            // Because the inner writer doesn't have native vectored write
+            // support, we should take care of buffering together the individual
+            // incoming bufs, even if the *total* length is larger than our
+            // buffer. We only want to skip our buffer if an *individual* write
+            // exceeds our buffer capacity.
+            let mut total_buffered = 0;
+
+            for buf in bufs {
+                if total_buffered == 0 {
```

Comment:
I have [rewritten](https://github.com/rust-lang/rust/blob/dc9493da4bce0f6af9ebeed30d9777102e2492dc/library/std/src/io/buffered/bufwriter.rs#L346) the implementation in #78768 to unroll the first iteration over a non-empty slice, which is the only one where flushing and potential write-through are possible. So this branch in the loop body is eliminated, and some of the checks that are unnecessary for the first slice are avoided.

Reply:
I'll check out the generated assembly; my expectation is that the optimizer can correctly take care of this (that is, it can reflow the control flow here ensure that the branch isn't re-checked on subsequent iterations).

Reply:
In this case, I prefer the simpler (non-duplicated) code and am willing to trust the optimizer to cover this case.